### DISCUSSION TITLE
코드의 출력 결과를 말해주세요.

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#54] [스코프]
- 코드의 출력 결과를 말해주세요.
   ```js
    function foo() {
      var x = 'local';
      console.log(x);
      return x;
    }
    
    foo();
    console.log(x);
   ```
  - 호이스팅은 어느 시점에 일어나는 지? (두번째 출력 결과가 undefined가 아닌 이유)

**Comment by publdaze** [comment]
> - 코드의 출력 결과를 말해주세요.
>    ```js
>     function foo() {
>       var x = 'local';
>       console.log(x);
>       return x;
>     }
>     
>     foo();
>     console.log(x);
>    ```

- foo함수가 먼저 호출되는데, x는 지역 변수로 첫 번째 출력에서는 'local'이 출력되고,
- 2번째 출력에서는 자신과 상위 스코프에 x가 선언되어 있지 않기 때문에 reference에러 발생

> - 호이스팅은 어느 시점에 일어나는 지?
- 함수가 호출된 후 함수 내부 코드가 실행되기 이전 단계에서 변수 선언이 일어남.
- 전역 변수의 경우 런타임 이전에 일어남.

**Comment by pipisebastian** [comment]
> 코드의 출력 결과를 말해주세요.
```tsx
 function foo() {
   var x = 'local';
   console.log(x);
   return x;
 }
 foo();
 console.log(x);
```
- 3 line 출력의 경우, 'local'이 출력되며, 8 line 출력의 경우, 참조할 x가 없기 때문에 reference error를 띄우게 됩니다. 이는 foo() 내부 x가 지역스코프이기 때문에 참조할 수 없습니다.

> 호이스팅은 어느 시점에 일어나는 지? (두번째 출력 결과가 undefined가 아닌 이유)

-  함수가 호출되고, 호이스팅이 일어나 변수가 선언됩니다.
